### PR TITLE
Make label action work for first time contributors

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

 

#2136 shows that the label workflow won't work for first time contributors due to Actions permissions.  Using `pull_request_target` makes the label use the PR base branch rather than the PR merged in.

### Details

- < feature1 or bug1 description >
- < feature2 or bug2 description >

### CODE REVIEW

**Comment by banesullivan** [comment]
`pull_request_target` is the right choice here however I think no matter what, forked contributions cannot use the GitHub token to label/comment/etc. See https://github.com/pyvista/pyvista/pull/2070#issuecomment-1028833556



| Scope | Default access(permissive) | Default access(restricted) | Maximum accessby forked repos |
|-- | -- | -- | -- |
| pull-requests | read/write | none | read |

**Review by banesullivan — APPROVED** [review]

**Review by banesullivan — CHANGES_REQUESTED** [review]
Actually, no. This will not work due to the conditions I set up in this workflow

File: .github/workflows/label.yml
Comment (banesullivan) [inline .github/workflows/label.yml:3]: The `if` conditions (see line 19) requires this to be run on `pull_request` events so that we can properly detect the head branch of the fork itself.

**Comment by banesullivan** [comment]
In order for this to not fail on forked contributions, we need a similar check to what is done in `.github/workflows/docker-package.yml`:

https://github.com/pyvista/pyvista/blob/eed641ebf0d80dee16d773750904d60afe29e59e/.github/workflows/docker-package.yml#L20

**Comment by banesullivan** [comment]
We should continue to use the `pull_request` even trigger and add a job-level `if` statement to prevent this from running on forked contributions.

@MatthewFlamm, AFAIK, `pull_request_target` vs `pull_request` has nothing to do with permissions but which version of the code is actually checked out. While intuitively it makes sense that `pull_request_target` = our version of the code so there should be a permissive token, there are actually security vulnerabilities in that case for forked contributions, so GitHub only issues `read`-access tokens to any forked contribution.

**Comment by MatthewFlamm** [comment]
> ~ I think~ no matter what, forked contributions cannot use the GitHub token to label/comment/etc. See https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token

I think this table applies only for `pull_request` although it is not explicit here.

This worked for automatic greetings, which is similar in terms of need for read/write access.  See #1838.  These were failing for first time contributors before using `pull_request_target`

See this part of the docs that explicitly state:

>Warning: For workflows that are triggered by the pull_request_target event, the GITHUB_TOKEN is granted read/write repository permission unless the permissions key is specified and the workflow can access secrets, even when it is triggered from a fork. 

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target

I'm okay if we close this based on security concerns too (and then should also remove the greetings action to remove the security concern there too.).  It is bad contributor experience to have a bunch of failed checks when opening a first PR.  If this is not implemented, is it possible to check the permissions of the github token before running?

File: .github/workflows/label.yml
Comment (MatthewFlamm) [reply .github/workflows/label.yml:3]: These can be changed once we agree on whether we should do this first or not.

**Comment by MatthewFlamm** [comment]
Sorry I didn't read all of your messages first.  I can see the value of keeping this label action running from forks in case any regular contributor prefers to work in their fork for whatever reason, but still uses the ci naming structure.  But having said that, it is no great loss to have to manually label those PR's in those cases.  If you prefer your suggestion, I'm fine closing this.

**Comment by banesullivan** [comment]
> This worked for automatic greetings, which is similar in terms of need for read/write access. See https://github.com/pyvista/pyvista/pull/1838. These were failing for first time contributors before using pull_request_target

That's a good point... I'm going to play with this in a separate repo (to not overload our CI) to see if I can get this to work well for forks

**Comment by MatthewFlamm** [comment]
> I'm going to play with this in a separate repo (to not overload our CI) to see if I can get this to work well for forks

I'll mark this as draft in the meantime.

**Comment by banesullivan** [comment]
@MatthewFlamm, thanks for looking into this and finding that `pull_request_target` does indeed yield a write access token

**Comment by MatthewFlamm** [comment]
I think we still need to update the if statements.

**Comment by MatthewFlamm** [comment]
Maybe it was a GitHub glitch but this PR was not labeled

**Comment by banesullivan** [comment]
It was not labeled because this action was changed `pull_request_target` - that means the code and thus the action YML is checked out from the target branch (`main`). The state of this workflow was `on: pull_request` for `main`, so it could never be executed for this PR as the target branch expected the PR branch to handle it and the PR branch expected the target branch to handle it

**Comment by adeak** [comment]
We still need to manually enable CI for first-time contributors, right?

**Comment by banesullivan** [comment]
@adeak, I recently set it so only "new users to GitHub" have to get approval

Screen Shot 2022-02-04 at 1 28 20 PM

**Comment by adeak** [comment]
I see, thanks @banesullivan. This means we'll have to be low-key vigilant with potential malicious PRs. But I'm still subscribed to all new PRs. Hmm, question is whether I'd be able to kick out actions from malicious PRs. Is my hunch correct that this is also restricted to the maintainers team now?

**Comment by adeak** [comment]
Nevermind, I can just check. I opened the repo level "actions" tab and I got a "delete workflow run" option in a given run's context menu. Sounds like I would be able to do that.

**Comment by banesullivan** [comment]
I'll try to keep an out for those too. if we see any come through, I'll immediately revert that setting, but for now, I have confidence in GitHub and want to make sure any new contributor and get going as soon as easily as possible. 

If you think we should steer on the side of caution, I can change that setting back to having us approve actions. I'm not too concerned, because I have seen many malicious PR personally

I think you're right that all of this control is limited to the "maintainers" team which I'm happy to add you back to 😃, just say the word!

**Comment by adeak** [comment]
I think the new setting strikes a nice balance between caution and ease of use. I just like to be clear on the (unlikely) possibilities to have a plan when something goes off the rails :)

**Pull request merged**
